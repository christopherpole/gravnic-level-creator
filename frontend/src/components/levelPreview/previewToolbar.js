import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';

import Button from '../common/button';
import { editLevel, restartLevel, undoMove } from '../../actions/levelPreview';

export const Wrapper = styled.div`
  grid-column: 1 / 2;
  margin-top: ${props => props.theme.structureSpacing};
`;

export const Toolbar = styled.ul`
  display: flex;
  justify-content: center;
  padding: 0;
`;

export const ActionContainer = styled.li`
  list-style-type: none;
  margin-right: calc(${props => props.theme.structureSpacing} / 2);

  &:last-child {
    margin-right: 0;
  }
`;

export const PreviewToolbar = ({
  editLevelAction,
  restartLevelAction,
  undoMoveAction,
  gameHistory,
}) => (
  <Wrapper id="editor-toolbar">
    <Toolbar>
      <ActionContainer>
        <Button
          id="btn-edit"
          onClick={() => {
            editLevelAction();
          }}
        >
          Edit
        </Button>
      </ActionContainer>
      <ActionContainer>
        <Button
          id="btn-restart"
          onClick={() => {
            restartLevelAction();
          }}
        >
          Restart
        </Button>
      </ActionContainer>
      <ActionContainer>
        <Button
          disabled={gameHistory.length <= 1}
          id="btn-undo"
          onClick={() => {
            undoMoveAction();
          }}
        >
          Undo
        </Button>
      </ActionContainer>
    </Toolbar>
  </Wrapper>
);

PreviewToolbar.propTypes = {
  editLevelAction: PropTypes.func.isRequired,
  restartLevelAction: PropTypes.func.isRequired,
  undoMoveAction: PropTypes.func.isRequired,
  gameHistory: PropTypes.array.isRequired,
};

const mapStateToProps = state => ({
  gameHistory: state.levelPreview.gameHistory,
});

const mapDispatchToProps = dispatch => ({
  editLevelAction: bindActionCreators(editLevel, dispatch),
  restartLevelAction: bindActionCreators(restartLevel, dispatch),
  undoMoveAction: bindActionCreators(undoMove, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(PreviewToolbar);
