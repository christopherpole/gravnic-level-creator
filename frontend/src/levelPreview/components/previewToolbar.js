import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';

import Button from 'common/button';
import { editLevel, restartLevel, undoMove, setFastMode } from '../actions';
import { getLevelPreviewButtonDisabledStates } from '../selectors';

export const Wrapper = styled.div`
  grid-column: 1 / 2;
  margin-top: ${props => props.theme.structureSpacing};
`;

export const Toolbar = styled.ul`
  display: flex;
  justify-content: center;
  padding: 0;
`;

export const ActionContainer = styled.li`
  list-style-type: none;
  margin-right: calc(${props => props.theme.structureSpacing} / 2);

  &:last-child {
    margin-right: 0;
  }
`;

export const PreviewToolbar = ({
  editLevelAction,
  restartLevelAction,
  undoMoveAction,
  setFastModeAction,
  buttonDisabledStates,
  fastMode,
}) => (
  <Wrapper id="editor-toolbar">
    <Toolbar>
      <ActionContainer>
        <Button
          id="btn-edit"
          onClick={() => {
            editLevelAction();
          }}
        >
          Edit
        </Button>
      </ActionContainer>
      <ActionContainer>
        <Button
          disabled={buttonDisabledStates.btnRestart}
          id="btn-restart"
          onClick={() => {
            restartLevelAction();
          }}
        >
          Restart
        </Button>
      </ActionContainer>
      <ActionContainer>
        <Button
          disabled={buttonDisabledStates.btnUndo}
          id="btn-undo"
          onClick={() => {
            undoMoveAction();
          }}
        >
          Undo
        </Button>
      </ActionContainer>
      <ActionContainer>
        <Button
          id="btn-set-fast-mode"
          onClick={() => {
            setFastModeAction(!fastMode);
          }}
        >
          Speed: {fastMode ? 'FAST' : 'NORMAL'}
        </Button>
      </ActionContainer>
    </Toolbar>
  </Wrapper>
);

PreviewToolbar.propTypes = {
  editLevelAction: PropTypes.func.isRequired,
  restartLevelAction: PropTypes.func.isRequired,
  undoMoveAction: PropTypes.func.isRequired,
  buttonDisabledStates: PropTypes.object.isRequired,
  setFastModeAction: PropTypes.func.isRequired,
  fastMode: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  fastMode: state.levelPreview.fastMode,
  buttonDisabledStates: getLevelPreviewButtonDisabledStates(state),
});

const mapDispatchToProps = dispatch => ({
  editLevelAction: bindActionCreators(editLevel, dispatch),
  restartLevelAction: bindActionCreators(restartLevel, dispatch),
  undoMoveAction: bindActionCreators(undoMove, dispatch),
  setFastModeAction: bindActionCreators(setFastMode, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(PreviewToolbar);
