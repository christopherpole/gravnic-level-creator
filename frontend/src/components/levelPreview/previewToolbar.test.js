import React from 'react';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { spy } from 'sinon';

import { PreviewToolbar } from './previewToolbar';

configure({ adapter: new Adapter() });

describe('The preview toolbar', () => {
  let props;

  beforeEach(() => {
    props = {
      editLevelAction: () => {},
      restartLevelAction: () => {},
      undoMoveAction: () => {},
      gameHistory: [[[1, 2, 3]], [[4, 5, 6]]],
    };
  });

  it('Renders without exploding', () => {
    const previewToolbar = shallow(<PreviewToolbar {...props} />);

    expect(previewToolbar).toHaveLength(1);
  });

  it('Matches the current snapshot', () => {
    const previewToolbar = shallow(<PreviewToolbar {...props} />);

    expect(toJson(previewToolbar)).toMatchSnapshot();
  });

  it('Matches the current snapshot when there are no moves to be undone', () => {
    const previewToolbar = shallow(<PreviewToolbar {...props} gameHistory={[[[1, 2, 3]]]} />);

    expect(toJson(previewToolbar)).toMatchSnapshot();
  });

  it('Fires the "editLevel" action when the edit level button is clicked', () => {
    const editLevelSpy = spy();
    const previewToolbar = shallow(<PreviewToolbar {...props} editLevelAction={editLevelSpy} />);
    const btnEditLevel = previewToolbar.find('#btn-edit');

    expect(editLevelSpy.calledOnce).toBe(false);
    btnEditLevel.simulate('click');
    expect(editLevelSpy.calledOnce).toBe(true);
  });

  it('Fires the "restartLevel" action when the restart level button is clicked', () => {
    const restartLevelSpy = spy();
    const previewToolbar = shallow(
      <PreviewToolbar {...props} restartLevelAction={restartLevelSpy} />,
    );
    const btnRestartLevel = previewToolbar.find('#btn-restart');

    expect(restartLevelSpy.calledOnce).toBe(false);
    btnRestartLevel.simulate('click');
    expect(restartLevelSpy.calledOnce).toBe(true);
  });

  it('Fires the "undoMove" action when the undo move button is clicked', () => {
    const undoMoveSpy = spy();
    const previewToolbar = shallow(<PreviewToolbar {...props} undoMoveAction={undoMoveSpy} />);
    const btnUndoMove = previewToolbar.find('#btn-undo');

    expect(undoMoveSpy.calledOnce).toBe(false);
    btnUndoMove.simulate('click');
    expect(undoMoveSpy.calledOnce).toBe(true);
  });
});
