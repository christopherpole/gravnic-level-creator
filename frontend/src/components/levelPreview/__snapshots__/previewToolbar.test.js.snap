// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The preview toolbar Matches the current snapshot 1`] = `
<styled.div
  id="editor-toolbar"
>
  <styled.ul>
    <styled.li>
      <Button
        disabled={false}
        id="btn-edit"
        onClick={[Function]}
      >
        Edit
      </Button>
    </styled.li>
    <styled.li>
      <Button
        disabled={false}
        id="btn-restart"
        onClick={[Function]}
      >
        Restart
      </Button>
    </styled.li>
    <styled.li>
      <Button
        disabled={false}
        id="btn-undo"
        onClick={[Function]}
      >
        Undo
      </Button>
    </styled.li>
  </styled.ul>
</styled.div>
`;

exports[`The preview toolbar Matches the current snapshot when there are no moves to be undone 1`] = `
<styled.div
  id="editor-toolbar"
>
  <styled.ul>
    <styled.li>
      <Button
        disabled={false}
        id="btn-edit"
        onClick={[Function]}
      >
        Edit
      </Button>
    </styled.li>
    <styled.li>
      <Button
        disabled={false}
        id="btn-restart"
        onClick={[Function]}
      >
        Restart
      </Button>
    </styled.li>
    <styled.li>
      <Button
        disabled={true}
        id="btn-undo"
        onClick={[Function]}
      >
        Undo
      </Button>
    </styled.li>
  </styled.ul>
</styled.div>
`;
